pyro/contrib/epidemiology/compartmental.py

```diff
@@ -0,0 +1,427 @@
+                                    dim=-1 - self.max_plate_nesting)
+
+        # Sample discrete auxiliary variables conditioned on the continuous
+        # variables sampled by _vectorized_model. This samples only time steps
+        # [0:duration]. Here infer_discrete runs a forward-filter
+        # backward-sample algorithm.
+        logger.info("Predicting latent variables for {} time steps..."
+                    .format(self.duration))
+        model = self._sequential_model
+        model = poutine.condition(model, samples)
+        model = particle_plate(model)
+        if self._dct is not None:
+            # Apply the same reparameterizer as during inference.
+            rep = DiscreteCosineReparam(smooth=self._dct)
+            model = poutine.reparam(model, {"auxiliary": rep})
+        model = infer_discrete(model, first_available_dim=-2 - self.max_plate_nesting)
+        trace = poutine.trace(model).get_trace()
+        samples = OrderedDict((name, site["value"])
+                              for name, site in trace.nodes.items()
+                              if site["type"] == "sample")
+
+        # Optionally forecast with the forward _generative_model. This samples
+        # time steps [duration:duration+forecast].
+        if forecast:
+            logger.info("Forecasting {} steps ahead...".format(forecast))
+            model = self._generative_model
+            model = poutine.condition(model, samples)
+            model = particle_plate(model)
+            trace = poutine.trace(model).get_trace(forecast)
+            samples = OrderedDict((name, site["value"])
+                                  for name, site in trace.nodes.items()
+                                  if site["type"] == "sample")
+
+        self._concat_series(samples, forecast)
+        return samples
+
+    # Internal helpers ########################################
+
+    def _concat_series(self, samples, forecast=0):
+        """
+        Concatenate sequential time series into tensors, in-place.
+
+        :param dict samples: A dictionary of samples.
+        """
+        for name in self.compartments + self.series:
+            pattern = name + "_[0-9]+"
+            series = []
+            for key in list(samples):
+                if re.match(pattern, key):
+                    series.append(samples.pop(key))
+            if series:
+                assert len(series) == self.duration + forecast
+                series = torch.broadcast_tensors(*series)
+                samples[name] = torch.stack(series, dim=-1)
+
+    def _generative_model(self, forecast=0):
+        """
+        Forward generative model used for simulation and forecasting.
+        """
+        # Sample global parameters.
+        params = self.global_model()
+
+        # Sample initial values.
+        state = self.initialize(params)
+        state = {i: torch.tensor(float(value)) for i, value in state.items()}
```

Comment:
what is this for exactly?

Reply:
`self.initialize` currently returns deterministic initial states. Because those states are used in `torch.nn.functional.pad` they must be Python scalars. However in this function those initial states need to start out as tensors.

The `state` dict in this function acts similarly to the `state` dict in SMCFilter: it is a framework-managed storage location where users can read and write values. Note the `self.transition_fwd(params, state, t)` call below updates this state dict in-place.